### PR TITLE
Unittesting implementation

### PR DESCRIPTION
**Pull request by nicolbidone**
image

### CODE REVIEW

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt [hunk 1]
```diff
@@ -28,18 +27,19 @@ open class MainActivity : BaseRxActivity() {
         setContentView(R.layout.activity_main)
 
         fab_download.setOnClickListener {
+            realmDestroy()
             presenter.requestGetCharacters()
         }
 
         fab_getStored.setOnClickListener {
             presenter.requestStoredCharacters()
+            realmDestroy()
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt]: why do you destroy de DB??

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/SetCharacterStoredUseCaseTest.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+class SetCharacterStoredUseCaseTest {
+
+    private val itemsCharacters = CharactersFactory.getMockCharacter()
+
+    private lateinit var characterStored: CharacterStored
+
+    @Before
+    fun setUp() {
+        characterStored = mock(CharacterStored::class.java)
+        `when`(characterStored.setCharacters(itemsCharacters)).then {}
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/SetCharacterStoredUseCaseTest.kt:22]: why do you have the "{}" ?

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    private var characterService = Mockito.mock(CharacterServices::class.java)
+
+    private lateinit var model: FragmentContracts.Model
+    private lateinit var getSingleCharacterServiceUseCase: GetSingleCharacterServiceUseCase
+
+    @Before
+    fun setUp() {
+        RxAndroidPlugins.setInitMainThreadSchedulerHandler { scheduler -> Schedulers.trampoline() }
+        getSingleCharacterServiceUseCase = GetSingleCharacterServiceUseCase(characterService, ZERO_VALUE)
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt]: this must be a mock

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        model = FragmentModel(getSingleCharacterServiceUseCase)
+    }
+
+    @Test
+    fun isOkGetSingleCharacterServiceUseCase() {
+        val itemsCharacters = CharactersFactory.getMockCharacter()
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt]: you have to mock a character here

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentPresenterTest.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+class FragmentPresenterTest {
+
+    private var view = mock(FragmentContracts.View::class.java)
+    private var characterService = mock(CharacterServices::class.java)
+    private var charactersFragment = mock(CharacterFragment::class.java)
+
+    private lateinit var model: FragmentContracts.Model
+    private lateinit var presenter: FragmentContracts.Presenter
+    private lateinit var getSingleCharacterServiceUseCase: GetSingleCharacterServiceUseCase
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentPresenterTest.kt]: this must be a mock

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentPresenterTest.kt [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            RxAndroidPlugins.setInitMainThreadSchedulerHandler { scheduler -> immediate }
+        }
+
+        private const val SIZE = TEN_VALUE
+    }
+
+    @Before
+    fun setUp() {
+        RxAndroidPlugins.setInitMainThreadSchedulerHandler { scheduler -> Schedulers.trampoline() }
+        getSingleCharacterServiceUseCase = GetSingleCharacterServiceUseCase(characterService, ZERO_VALUE)
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentPresenterTest.kt]: this must be a mock

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+class CharacterModelTests {
+
+    private var characterServiceImp = Mockito.mock(CharacterServices::class.java)
+    private var characterStoredImp = Mockito.mock(CharacterStored::class.java)
+
+    private lateinit var model: CharacterContracts.Model
+    private lateinit var getCharacterServiceUseCase: GetCharacterServiceUseCase
+    private lateinit var getCharacterStoredUseCase: GetCharacterStoredUseCase
+    private lateinit var setCharacterStoredUseCase: SetCharacterStoredUseCase
+
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: the UseCases must be mocks

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        model = CharacterModel(getCharacterServiceUseCase, getCharacterStoredUseCase, setCharacterStoredUseCase)
+
+    }
+
+    @Test
+    fun isOkGetCharacterDataServiceUseCase() {
+        val itemsCharacters = CharactersFactory.getMockCharacter()
+        val observable = Observable.just(itemsCharacters)
+        Mockito.`when`(getCharacterServiceUseCase()).thenReturn(observable)
+    }
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: You forget the verify here.

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Test
+    fun isOkGetCharacterStoredUseCase() {
+        val itemsCharacters = CharactersFactory.getMockCharacter()
+        Mockito.`when`(getCharacterStoredUseCase()).thenReturn(itemsCharacters)
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: You forget the verify here.

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Test
+    fun isOkSetCharacterStoredUseCase() {
+        val itemsCharacters = CharactersFactory.getMockCharacter()
+        Mockito.`when`(setCharacterStoredUseCase(itemsCharacters)).then { }
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: You forget the verify here.

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt [hunk 11]
```diff
@@ -1,80 +1,151 @@
...
-    private lateinit var getCharacterServiceUseCase: GetCharacterServiceUseCase
+    private var view = mock(CharacterContracts.View::class.java)
+    private var characterService = mock(CharacterServices::class.java)
+    private var characterStored = mock(CharacterStored::class.java)
 
+    private lateinit var model: CharacterContracts.Model
+    private lateinit var presenter: CharacterContracts.Presenter
+    private lateinit var getCharacterServiceUseCase: GetCharacterServiceUseCase
+    private lateinit var getCharacterStoredUseCase: GetCharacterStoredUseCase
+    private lateinit var setCharacterStoredUseCase: SetCharacterStoredUseCase
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt:38]: the UseCases must be mocks

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt (hunk 3 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt (hunk 4 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentModelTest.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentPresenterTest.kt (hunk 5 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/fragment/mvp/FragmentPresenterTest.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt (hunk 7 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt (hunk 8 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt (hunk 9 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt (hunk 10 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModelTests.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt (hunk 11 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt:38]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/SetCharacterStoredUseCaseTest.kt (hunk 2 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/SetCharacterStoredUseCaseTest.kt:22]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt (hunk 1 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt]: Done

**Review by SartiMau — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Pull request merged**